compiler/rustc_lint/messages.ftl

```diff
@@ -440,17 +440,19 @@ lint_non_local_definitions_cargo_update = the {$macro_kind} `{$macro_name}` may
 
 lint_non_local_definitions_deprecation = this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
 
-lint_non_local_definitions_impl = non-local `impl` definition, they should be avoided as they go against expectation
+lint_non_local_definitions_impl = non-local `impl` definition, `impl` blocks should be written at the same level as their item
     .help =
-        move this `impl` block outside the of the current {$body_kind_descr} {$depth ->
+        move this `impl` block and outside the of the current {$body_kind_descr} {$depth ->
             [one] `{$body_name}`
            *[other] `{$body_name}` and up {$depth} bodies
         }
-    .non_local = an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
-    .exception = one exception to the rule are anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type
+    .without_trait = methods and assoc const are still usable outside the current expression, only `impl Local` and `impl dyn Local` are local and only if the `Local` type is at the same nesting as the `impl` block
+    .with_trait = an `impl` is never scoped, even when it is nested inside an item, as it may impact type checking outside of that item, which can be the case if neither the trait or the self type are at the same nesting level as the `impl`
+    .bounds = `impl` may be usable in bounds, etc. from outside the expression, which might e.g. make something constructible that previously wasn't, because it's still on a publicly-visible type
+    .exception = anon-const (`const _: () = {"{"} ... {"}"}`) at top-level module and anon-const at the same nesting as the trait or type are consider to be transparent regarding the nesting level
```

Comment:
what does "transparent regarding the nesting level" mean?

Reply:
it means that putting a non-local `impl` def inside a `const _` will make it "global"
```rust
trait Trait {}
struct Test;

const _: () = {
    impl Trait for Test {} // for the lint Trait, Test and the `impl` are at the same nesting
                           // even through they are clearly not
};
```

Reply:
Interesting. I think it might want to say something more like
```suggestion
    .exception = items in an anon const (`const _: () = {"{"} ... {"}"}`) are treated as in the same scope as the anon const's declaration
```

Reply:
Mostly "transparent" seems slightly... opaque, ironically, when we want to say "this level of nesting is ignored" or "these two scopes are linted as if the same scope"

Reply:
Simpler, I like it. Updated.